Add new 'config' syntax

jordalgo opened the pull request:
This new feature allows users to set environment variables and other configuration in bpftrace scripts. This makes scripts more portable e.g. it's easier to reproduce behaviors without having to remember how scripts were invoked.

Environment variables passed on the command line take precedence
over ones set in the script itself.

Example:
```
sudo bpftrace -v -e 'config = { BPFTRACE_STACK_MODE=raw; } kprobe:vfs_read {
print(kstack()); }'
```


##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

jordalgo commented:
I'm sure there are a bunch of things I didn't get right in this implementation but creating an RFC for the language syntax. I'll also be sure to add tests later on in the process.

cc @ajor

viktormalik commented:
First of all, I think that this is a very useful feature which can improve portability of bpftrace scripts, thanks!

I didn't review the code at all, but have some thoughts on the syntax. Similarly to what @ajor [describes on iterators syntax](https://github.com/iovisor/bpftrace/issues/2775#issuecomment-1779899452), this breaks the traditional event-based concept of bpftrace probes.

I'd rather see initializing env variables in the `BEGIN` block (similarly to what awk does). To do that, we could introduce new syntax for environment variables, e.g. (inspired by CMake, we can come with something else):
```
BEGIN
{
    $ENV{BPFTRACE_STACK_MODE} = "raw";
}
```
The advantage of this is that we could eventually allow usage of environment variables inside other probes, too (and the value would be replaced by actual value upon compile time - similarly to positional parameters).

jordalgo commented:
> this breaks the traditional event-based concept of bpftrace probes. I'd rather see initializing env variables in the BEGIN block

Yeah, I considered this too. I think I didn't do that initially because I worried about expectations around ordering in terms of when env variables get evaluated and set (in relation to evaluation of other things in `BEGIN`) but maybe that's not a concern (at least until we start allowing them to be used in place of other r-values in other probes). Anyway, I'm good with moving them to `BEGIN`.

As far as the syntax:
```
$ENV{BPFTRACE_STACK_MODE} = "raw";
```

Are you thinking we need `$ENV` prefix as opposed to just using `BPFTRACE_` so that we can support user defined env variables in the future? If so, what is the difference between them and global/scratch variables? Or is it just so we don't have to always prefix env variables with `BPFTRACE_`?

viktormalik commented:
> Are you thinking we need `$ENV` prefix as opposed to just using `BPFTRACE_` so that we can support user defined env variables in the future? If so, what is the difference between them and global/scratch variables? Or is it just so we don't have to always prefix env variables with `BPFTRACE_`?

I'd say that they are quite different, both from user and from bpftrace point of view. As a user, I prefer to know which variables come from the environment and may therefore have unexpected values. As a bpftrace developer, it feels like parsing and handling env variables are different from normal variables (e.g. they can be used before their first assignment) and the prefix should simplify our work. Ofc if we stick with bpftrace-specific variables only (having the `BPFTRACE_` prefix), there's no need to add the `ENV` prefix.

fbs commented:
I think putting everything in BEGIN will make BEGIN really confusing. #1945 has a proposal for CLI parsing, which would then also have to go in begin. 

If you go for something like `bpftrace:env:stackmode { "blaat" }`, then you dont need any special parsing tricks. Its just a standard probe, with a single value in it. IIRC the AST accepts that without modification. 
Because it starts with `bpftrace:` we can treat it as a special probe, with its own AST pass, with very little modifications outside the pass itself, we just need to filter it from other passes (could even do an ast rewrite for that).
I think it will also stick to the 'event based' probe idea, although its a special probe that fires once

jordalgo commented:
> I think putting everything in BEGIN will make BEGIN really confusing.

Yeah, I'm starting to agree. I just don't like the inter-splicing of env var assignments with anything else in BEGIN which might imply some ordering when there is none; env vars will always get set before we do any codegen.

> `bpftrace:env:stackmode { "blaat" }`

This syntax feels a little confusing to me. I'd rather it look more like the actual env variable setting on the command line e.g. `env:BPFTRACE_STACK_MODE { perf }` or `bpftrace:env:BPFTRACE_STACK_MODE { perf }`  but TBH, I'm not crazy about the floating r-value in the brackets and prefer the original proposal `ENV { BPFTRACE_STACK_MODE=perf }`, which, if you look at the code in this PR, is not that invasive in terms of modifications - most of the business logic is in semantic_analyzer.cpp (though I'm not sure if that's the best place).

viktormalik commented:
@fbs has a good point that the syntax should be unified with argument parsing and having everything in `BEGIN` would make it too complicated. I kind of start liking the idea of "meta" probes starting with `bpftrace:` prefix which are not executed but instead hold some script metadata. They seem quite intuitive, consistent with the current language, and quite extensible for future.

I agree with @jordalgo that we should use the original name - either `bpftrace:env:BPFTRACE_STACK_MODE { perf }` or, even better from my point of view,  `bpftrace:env { BPFTRACE_STACK_MODE = perf }`.

FWIW, I put this on the agenda of the next office hours meeting.

viktormalik commented:
Also note that this has a big overlap with #728.

fbs commented:
> > `bpftrace:env:stackmode { "blaat" }`
> 
> This syntax feels a little confusing to me. I'd rather it look more like the actual env variable setting on the command line e.g. `env:BPFTRACE_STACK_MODE { perf }` or `bpftrace:env:BPFTRACE_STACK_MODE { perf }` but TBH, I'm not crazy about the floating r-value in the brackets and prefer the original proposal `ENV { BPFTRACE_STACK_MODE=perf }`, which, if you look at the code in this PR, is not that invasive in terms of modifications - most of the business logic is in semantic_analyzer.cpp (though I'm not sure if that's the best place).

My main issue with this is that you now have a 3rd way to assign variables. So there will be `$var =x`, `@map = x` and `SPECIAL = x`. All the AST changes are due to this.
I agree that the `bpftraceprobe { val }` syntax is a bit weird too, my idea is that it should be read like `bpftrace:probe { return val }`, so the probe fires and returns a value. This is also a bit flawed as there is no const evaluation into a value.
Maybe there is another option that avoids creating new syntax?

The reason for me to go for `stack_mode` instead of `BPFTRACE_STACK_MODE` is because the latter is just shell conventions. bpftrace_ because env vars are not namespaced, this avoids conflict. uppercase because that is also normal for the env, probably avoid confusion with shell local variables? Either way, I'm not sure we should bring that shell detail into  bpftrace. Maybe we should even drop env from the name, why not just `bpftrace:config:stack_mode`?

viktormalik commented:
> The reason for me to go for `stack_mode` instead of `BPFTRACE_STACK_MODE` is because the latter is just shell conventions. bpftrace_ because env vars are not namespaced, this avoids conflict. uppercase because that is also normal for the env, probably avoid confusion with shell local variables? Either way, I'm not sure we should bring that shell detail into bpftrace. Maybe we should even drop env from the name, why not just `bpftrace:config:stack_mode`?

If we define (and document) this as the idiomatic way of configuring bpftrace, then I'm all good with the last option. We can then just say that the configuration options can be also set via env variables which, by shell conventions, are capitalized and use the `BPFTRACE_` prefix.

jordalgo commented:
Do we think there would ever be a difference between the `bpftrace:config` variables and env vars that you can set when invoking bpftrace? If not, then I'm not sure why you'd want to create a new thing that users would have to remember. I understand that these are shell conventions but is it so bad that we adopt them specifically so users don't have to remember which one needs a prefix and uppercase spelling? Somewhat rhetorical - I don't think it's so bad :)

To avoid another way of setting variables, what if we just re-used `$` e.g. `$BPFTRACE_` - it could be an AST change or we could just add branching logic for that ident prefix in semantic_analyzer.cpp.

jordalgo commented:
Ok, I think this might be ready-ish. I'm sure there were a few things I probably missed.

viktormalik commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2672,6 +2672,222 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
+    {
+      invalidType();
+    }
+    else if (!env_is_set)
```
TBH, I don't like that this check needs to be done for every config option. I can see the problem here (we'd need to parse env vars between doing semantic analysis of config and of the rest of the script) but I'm wondering if we can come up with some more clever and robust solution.

Idea: encapsulate config in a new class with various exported setters (for setting from env, script, cli, ...) which would automatically handle the priorities? It's a bit of work but I'm afraid that with the proposed approach, we will mess up the priorities sooner or later for some option.

jordalgo replied:
Yeah I don't like it either. Another class sounds like a good idea. Let me think on that in the context of @danobi's suggestion to move this parsing to a separate AST pass.

viktormalik commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2672,6 +2672,222 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
+    return false;
+  };
+
+  if (matches("BPFTRACE_STACK_MODE"))
```
This is probably super-nit but since we want to encourage moving from env vars to config options, we should not use the env var name rather than the config name (i.e. `"stack_mode"`) here. Also, if we introduce new config options without corresponding env vars inn future, the code will remain consistent.

viktormalik commented on src/bpftrace.h:
```diff
@@ -202,12 +213,8 @@ class BPFtrace
-    per_program,
-    none,
-  } user_symbol_cache_type_;
+  bool verify_llvm_ir_ = false;
+  UserSymbolCacheType user_symbol_cache_type_;
```
It feels like introduction of these two fields could be moved into separate commit(s).

viktormalik commented on tests/runtime/config:
```diff
@@ -0,0 +1,25 @@
+NAME config as env var
```
not env var, right?
```suggestion
NAME config
```

jordalgo replied:
Maybe the name is confusing but I want to test that the env variable style e.g. `BPFTRACE_` works for setting the config (as opposed to the test below)

viktormalik replied:
Ah, right, I just misunderstood it. I think that the current name is ok.

viktormalik reviewed (commented):
Thanks for this work, I'm really looking forward to get it in! I've a couple of comments, though, see below.

Also, I'd love to see the work split into multiple commits, e.g. the refactoring of methods for setting various configs from env vars could be moved into separate commit(s).

danobi reviewed (commented):
overall looks reasonable; haven't had a chance to look at details yet though.

one high level comment: could we put the config parsing (and future cli parsing) into a new AST pass? For stuff like AOT and @ajor's language plans, it would be useful to be able to selectively disable config/cli parsing via `PassManager`

jordalgo commented:
> Also, I'd love to see the work split into multiple commits, e.g. the refactoring of methods for setting various configs from env vars could be moved into separate commit(s).

Good suggestion. Let me work on that.

jordalgo commented:
> could we put the config parsing (and future cli parsing) into a new AST pass?

Good call. Let me work on that too :)

viktormalik reviewed (commented):
Looks good to me, just one small nit.

Thanks for this work!

jordalgo commented on src/config.cpp:
```diff
@@ -34,16 +34,21 @@ Config::Config(bool has_cmd, bool bt_verbose) : bt_verbose_(bt_verbose)
     return true;
   }
-  else if (prevSource == ConfigSource::env_var)
+  else if (prevSource == source || prevSource == ConfigSource::env_var)
   {
     return false;
   }
+  else if (prevSource == ConfigSource::script &&
+           source == ConfigSource::env_var)
+  {
+    return true;
+  }
```
Good call, that's certainly cleaner. Will fix.

ajor commented on tests/config_analyser.cpp:
Also I think it would be good to check the error messages of failures, to ensure they stay reasonable. I started doing this in semantic analyser tests here: https://github.com/iovisor/bpftrace/blob/98960e19de63b670d753b77939dbdc78dff6ac15/tests/semantic_analyser.cpp#L2834

It'd be nice to start this early in this file before too many tests get added.

jordalgo replied:
Added a few of both kinds of tests

ajor commented on docs/reference_guide.md:
````diff
@@ -951,6 +955,30 @@ Attaching 1 probe...
+Example:
+```
+config = {
+    BPFTRACE_STACK_MODE=perf;
+    BPFTRACE_MAX_PROBES=2
````
```suggestion
    stack_mode=perf;
    max_probes=2
```

I'd prefer we make the non-env-var syntax the documented standard

ajor reviewed (changes requested):
By adding these to the language syntax we're making them a lot more permanent than the env-vars were. I haven't thought too hard about the names yet (most might be fine!) but just want to suggest that we all consider what they're called. e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?

On another note, I found some odd errors while playing around with this feature. Using non-literals results in a "none" type error:
```
$ sudo ./bpftrace -e 'config = { MAP_KEYS_MAX = $x } BEGIN { }'
stdin:1:25-26: ERROR: Invalid type for MAP_KEYS_MAX. Type: none                                                                                                       
config = { MAP_KEYS_MAX = $x } BEGIN { }                                                                                                                              
                        ~
```

Also for expressions which can be evaluated at compile-time:
```
$ sudo ./bpftrace -e 'config = { MAP_KEYS_MAX = 1 + 1 } BEGIN { }'                                           
stdin:1:25-26: ERROR: Invalid type for MAP_KEYS_MAX. Type: none                                                                                                       
config = { MAP_KEYS_MAX = 1 + 1 } BEGIN { }                                                                                                                           
                        ~
```

jordalgo commented:
> e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?

Yeah this bothered me too. Happy to change it (and cat_bytes_max) as there is a 3 to 2 majority for putting 'max' first. Should we just have backward support for BPFTRACE_MAP_KEYS_MAX or are we fine changing the env variable as well to BPFTRACE_MAX_MAP_KEYS (and MAX_CAT_BYTES) ?

> Using non-literals results in a "none" type error:

Good catch. Let me add a check for all non-literals. IMHO we shouldn't allow any other type (at least not in this first iteration).

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2685,6 +2682,23 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
+    auto &ty = assignTy.type;
+    if (ty == Type::none)
+    {
+      LOG(ERROR, assignment.loc, err_)
```
Nope. Going to delete. Good catch.

jordalgo commented:
Side question: Is there a reason why the language doesn't support boolean literals? Would be nice to do `debug_output=true` instead of `debug_output=1`

ajor commented:
> Side question: Is there a reason why the language doesn't support boolean literals? Would be nice to do `debug_output=true` instead of `debug_output=1`

No reason in particular, it just hasn't come up before. Might be good to add at some point though.

ajor commented on src/config.h:
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
 
+// These strings match the env variables (minus the 'BPFTRACE_' prefix)
+const std::map<std::string, ConfigKey> CONFIG_KEY_MAP = {
+  { "node_max", ConfigKeyInt::ast_max_nodes },
```
Similar to "verify_llvm_ir" - this one's just used for fuzzing: https://github.com/iovisor/bpftrace/pull/1601/commits/fb87a9f9f476af3d6cd6d3b42bea8bbd7235cc07
I'd say exclude.

>    This introduces `BPFTRACE_NODE_MAX` environment variable, and it
>    controls the maximum number of AST nodes. This feature is intended for
>    developers, and its usage is not included in the reference guide.

ajor commented on src/config.h:
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
+const std::map<std::string, ConfigKey> CONFIG_KEY_MAP = {
+  { "node_max", ConfigKeyInt::ast_max_nodes },
+  { "cat_bytes_max", ConfigKeyInt::cat_bytes_max },
+  { "debug_output", ConfigKeyBool::debug_output },
```
Also developer-only, so I'd say exclude

ajor commented:
> > e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?
> 
> Yeah this bothered me too. Happy to change it (and cat_bytes_max) as there is a 3 to 2 majority for putting 'max' first. Should we just have backward support for BPFTRACE_MAP_KEYS_MAX or are we fine changing the env variable as well to BPFTRACE_MAX_MAP_KEYS (and MAX_CAT_BYTES) ?

Yeah I'd vote for putting "max" first too. No strong opinion on changing the environment variables.

jordalgo commented on src/config.h:
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
+  { "max_probes", ConfigKeyInt::max_probes },
+  { "max_bpf_progs", ConfigKeyInt::max_bpf_progs },
+  { "max_type_res_iterations", ConfigKeyInt::max_type_res_iterations },
+  { "no_cpp_demangle", ConfigKeyBool::no_cpp_demangle },
```
Ugh yes! I'm not a fan of negative named flags.

jordalgo commented:
@ajor I put the config and env var name changes into a separate commit. Let me know if you want me to merge it back into the first.

jordalgo commented on src/main.cpp:
```diff
@@ -363,6 +363,42 @@ static std::optional<struct timespec> get_delta_taitime()
       return false;
   }
 
+  // Todo: delete these warnings in the future
+  if (std::getenv("BPFTRACE_NO_CPP_DEMANGLE"))
+  {
+    LOG(ERROR) << "BPFTRACE_NO_CPP_DEMANGLE is deprecated. Use "
+                  "BPFTRACE_CPP_DEMANGLE=0 instead.";
+    return false;
+  }
+
+  if (std::getenv("BPFTRACE_CAT_BYTES_MAX"))
+  {
+    LOG(ERROR) << "BPFTRACE_CAT_BYTES_MAX is deprecated. Use "
+                  "BPFTRACE_MAX_CAT_BYTES instead.";
+    return false;
+  }
+
+  if (std::getenv("BPFTRACE_MAP_KEYS_MAX"))
+  {
+    LOG(ERROR) << "BPFTRACE_MAP_KEYS_MAX is deprecated. Use "
+                  "BPFTRACE_MAX_MAP_KEYS instead.";
+    return false;
+  }
+
+  if (std::getenv("BPFTRACE_STRLEN"))
+  {
+    LOG(ERROR)
+        << "BPFTRACE_STRLEN is deprecated. Use BPFTRACE_MAX_STRLEN instead.";
+    return false;
+  }
+
+  if (std::getenv("BPFTRACE_NODE_MAX"))
+  {
+    LOG(ERROR) << "BPFTRACE_NODE_MAX is deprecated. Use BPFTRACE_MAX_AST_NODES "
+                  "instead.";
+    return false;
+  }
```
That works for me. No strong feelings on this subject but I do like Strobelight.

jordalgo commented on src/main.cpp:
```diff
@@ -448,6 +450,7 @@ ast::PassManager CreateDynamicPM()
 ast::PassManager CreateAotPM()
 {
   ast::PassManager pm;
+  pm.AddPass(ast::CreateConfigPass());
```
Glad you called this out. I'm not very familiar with the AOT stuff so fine to drop it until we figure out how it'll work.

jordalgo commented on src/parser.yy:
```diff
@@ -349,6 +360,12 @@ stmt_list:
+config_assign_stmt_list:
+                config_assign_stmt ";" config_assign_stmt_list { $$ = $3; $3->insert($3->begin(), $1); }
+        |       config_assign_stmt                             { $$ = new ast::StatementList; $$->push_back($1); }
+        |       %empty                                         { $$ = new ast::StatementList; }
```
Probably not very useful but bpftrace also supports empty BEGIN and END. No strong feeling either way.

danobi replied:
Hmm, empty begin and end can still be used to check that probes attach correctly. But I suppose that could be said for empty config block too.

No strong feeling either way

jordalgo replied:
Seems easy enough to get rid of. I'll try removing.

danobi reviewed (approved):
lgtm!

ajor commented on src/ast/passes/config_analyser.cpp:
```diff
@@ -0,0 +1,205 @@
+    return PassResult::Success();
+  };
+
+  return Pass("Semantic", fn);
```
```suggestion
  return Pass("ConfigAnalyser", fn);
```

jordalgo replied:
Goodc catch!

ajor reviewed (commented):
Just some copy-paste mistakes and nits left, looks good otherwise

jordalgo commented on src/main.cpp:
```diff
@@ -120,19 +121,19 @@ void usage()
   std::cerr << "    BPFTRACE_BTF                      [default: none] BTF file" << std::endl;
   std::cerr << "    BPFTRACE_CACHE_USER_SYMBOLS       [default: auto] enable user symbol cache" << std::endl;
-  std::cerr << "    BPFTRACE_CAT_BYTES_MAX            [default: 10k] maximum bytes read by cat builtin" << std::endl;
+  std::cerr << "    BPFTRACE_MAX_CAT_BYTES            [default: 10k] maximum bytes read by cat builtin" << std::endl;
```
Good call. Will do

jordalgo commented on docs/reference_guide.md:
```diff
@@ -170,19 +171,19 @@ OPTIONS:
     BPFTRACE_BTF                        [default: none] BTF file
     BPFTRACE_CACHE_USER_SYMBOLS         [default: (see docs)] enable user symbol cache
-    BPFTRACE_CAT_BYTES_MAX              [default: 10k] maximum bytes read by cat builtin
+    BPFTRACE_MAX_CAT_BYTES              [default: 10k] maximum bytes read by cat builtin
```
There was a PR floating around here (https://github.com/iovisor/bpftrace/pull/2754) to just use the man page and delete the reference guide but I think it got abandoned :(

jordalgo commented on src/main.cpp:
```diff
@@ -362,6 +370,35 @@ static std::optional<struct timespec> get_delta_taitime()
+      }))
+    return false;
+
+  if (!get_uint64_env_var("BPFTRACE_MAX_CAT_BYTES", [&](uint64_t x) {
```
Doh. Stupid autocomplete :) Good catch.

jordalgo commented:
@ajor Thanks for the sharp-eyed review. I think I fixed all that you mentioned.

Additionally, I added one more commit whereby I extract out all the env-only configs from the config class. There were many other environment variables that weren't included in the config (e.g. BPFTRACE_BTF, BPFTRACE_VMLINUX, etc...) so
it was either include them in the config class or extract out similar ones. I chose the latter as it makes more sense to only include config vars that can be set by both the env and the config syntax (less indirection and clearer delineation).

jordalgo commented:
@viktormalik @danobi @ajor - Thanks so much for the great reviews and guidance on this one!
